### PR TITLE
Build out Asset holder contract

### PR DESCRIPTION
**Pull request by geoknee**
This PR scaffolds out much of what we need for `ETHAssetHolder` and `ERC20AssetHolder` contracts. These contracts extend a base `AssetHolder` contract, and currently have `deposit` and `withdraw` methods, and storage for tracking `holdings` and `outcomeHashes`. The base contract defines a permission-ed `setOutcome` method, accessible only by calling `pushOutcome` on the `NitroAdjudicator` (a small extension of `ForceMove`).

This implementation largely follows the specification in `/docs/`.

- [x] Add Outcomes contract specifying various structs
- [x] Add base AssetHolder contracts (largely mocked up)
- [x] Add `isExternalAddress` function (test passes)
- [x] Tweak `_setOutcome` method
- [x] Implement `deposit`
- [x] Implement `withdraw`
- [x] Add ETH and ERC20 holders that extend AssetHolder
- [x] Add a simple Token is ERC20 contract which mints coins to the deployer
- [x] Modify deployment script so that `AssetHolder`s know address of `NitroAdjudicator` / `ForceMove` contract and `Token` contract
- [x] `NitroAdjudicator` extends `ForceMove` and has `pushOutcome` method
- [x] `pushOutcome` test (well-formed tx)
- [x] `pushOutcome` test (ill-formed tx)

---
Further work for separate PRS:

#30 :
- [ ] Test `deposit`
- [ ] Test `withdraw`

(A `deposit` test for ETHAssetHolder is included in this PR, but does not test the `Deposited` events)

TBC:
- [ ] Implement `transferAll` 
- [ ] implement `claimAll`
- [ ] test `transferAll`
- [ ] test `claimAll`

### CODE REVIEW

File: contracts/AssetHolder.sol [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        // Prevents replay attacks:
+        // It's required that the participant signs the message, meaning only
+        // the participant can authorize a withdrawal.
+        // Moreover, the participant should sign the address that they wish
+        // to send the transaction from, preventing any replay attack.
+        address participant; // the account used to sign commitment transitions
+        bytes32 destination; // either an account or a channel
+        uint256 amount;
+        address sender; // the account used to sign transactions
+    }
```
Comment (tomclose) [inline contracts/AssetHolder.sol]: Is this still required?

File: contracts/AssetHolder.sol [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    //     );
+
+    //     require(
+    //         amount <= channelAffordsForDestination,
+    //         "Transfer: channel cannot afford the requested transfer amount"
+    //     );
+
+    //     holdings[destination] = holdings[destination] + amount;
+    //     holdings[channel] = holdings[channel] - amount;
+    // }
```
Comment (tomclose) [inline contracts/AssetHolder.sol]: Can probably just delete this.

File: contracts/ERC20AssetHolder.sol [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    // **************
+    // Asset Management
+    // **************
+
+    function deposit(bytes32 destination, uint256 expectedHeld, uint256 amount) public {
+        require(_token.transferFrom(msg.sender, address(this), amount), 'Could not deposit ERC20s');
+
+        uint256 amountDeposited;
+        // This protects against a directly funded channel being defunded due to chain re-orgs,
```
Comment (tomclose) [inline contracts/ERC20AssetHolder.sol]: Maybe reword: this allows participants to reduce the wait between deposits, while protecting them from losing funds by depositing too early. Specifically it protects against the scenario:
1. Participant A deposits
2. Participant B sees A's deposit, which means it is now safe for them to deposit
3. Participant B submits their deposit
4. The chain re-orgs, leaving B's deposit in the chain but not A's

File: contracts/ERC20AssetHolder.sol [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            recoverSigner(abi.encode(authorization), _v, _r, _s) == participant,
+            'Withdraw: not authorized by participant'
+        );
+
+        holdings[_addressToBytes32(participant)] = holdings[_addressToBytes32(participant)].sub(
+            amount
+        );
+        // Decrease holdings before calling transfer (protect against reentrancy)
+        _token.transfer(destination, amount);
+    }
```
Comment (tomclose) [inline contracts/ERC20AssetHolder.sol]: I don't think we need `withdraw` anymore. The new `transferAll` method will automatically withdraw if the address is a participant address.

File: contracts/ETHAssetHolder.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        AdjudicatorAddress = _AdjudicatorAddress;
+    }
+
+    modifier AdjudicatorOnly {
+        require(msg.sender == AdjudicatorAddress, 'Only the NitroAdjudicator is authorized');
+        _;
+    }
+
+    // **************
+    // ETH and Token Management
```
Comment (tomclose) [inline contracts/ETHAssetHolder.sol]: Just "ETH Management"? Or do we even need this?

File: contracts/ETHAssetHolder.sol [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        }
+        holdings[destination] = holdings[destination].add(amountDeposited);
+        if (amountDeposited < amount) {
+            // refund whatever wasn't deposited.
+            msg.sender.transfer(amount - amountDeposited); // TODO use safeMath here
+        }
+        emit Deposited(destination, amountDeposited, holdings[destination]);
+    }
+
+    function withdraw(
```
Comment (tomclose) [inline contracts/ETHAssetHolder.sol]: Probably don't need `withdraw` (as above).

File: contracts/NitroAdjudicator.sol [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                            outcomeHash
+                        )
+                    )
+                ) ==
+                channelStorageHashes[channelId],
+            'Submitted data does not match storage'
+        );
+
+        // effects
+        AssetHolder(assetHolder).setOutcome(channelId, outcomeHash);
```
Comment (tomclose) [inline contracts/NitroAdjudicator.sol]: The `assetHolder` needs to be taken from the outcome. You'll need to parse the `outcome` into an array of format `(assetHolder, labelledOutcome)[]`, and then loop over this array calling `setOutcome(channelId, hash(labelledOutcome))`

File: contracts/Outcome.sol [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    // Outcome = AssetOutcome[]
+    // AssetOutome = (AssetID, AllocationOrGuarantee)
+    // AllocationOrGuarantee = Allocation | Guarantee
+    // Allocation = AllocationItem[]
+    // AllocationItem = (Destination, Amount)
+    // Guarantee = (ChannelAddress, Destination[])
+    // Destination = ChannelAddress | ExternalAddress
+
+    struct AssetOutcome {
+        address token;
```
Comment (tomclose) [inline contracts/Outcome.sol]: Should be `assetHolderAddress`.

File: test/ForceMove/_acceptableWhoSignedWhat.test.ts [hunk 9]
```diff
@@ -1,8 +1,8 @@
 import {ethers} from 'ethers';
 import {expectRevert} from 'magmo-devtools';
 // @ts-ignore
-import ForceMoveArtifact from '../../build/contracts/TESTForceMove.json';
-import {setupContracts} from './test-helpers';
+import ForceMoveArtifact from '../../build/contracts/TESTNitroAdjudicator.json';
```
Comment (tomclose) [inline test/ForceMove/_acceptableWhoSignedWhat.test.ts]: I think it might be worth continuing to test `ForceMove` directly. It's entirely possible that it will be wrapped in other adjudicators at a later date.

File: contracts/AssetHolder.sol (hunk 1 above)
Comment (geoknee) [reply contracts/AssetHolder.sol]: I think we probably still need the `participant`'s signature on something. But we may not need to include the `sender` in the message hash, if `withdraw` is to be an internal call from `transferAll`.

File: contracts/AssetHolder.sol (hunk 2 above)
Comment (geoknee) [reply contracts/AssetHolder.sol]: 👍

File: contracts/ERC20AssetHolder.sol (hunk 4 above)
Comment (geoknee) [reply contracts/ERC20AssetHolder.sol]: Ok -- but my preference is to leave this here until `transferAll` exists. That way, we may be able to support / build against directly funded channels before the full nitro functionality is available. Also, we may want to convert this to an internal method that `transferAll` can make use of (particularly in the ERC20 case). That's instead of nuking it now and reintroducing later.

File: contracts/ETHAssetHolder.sol (hunk 5 above)
Comment (geoknee) [reply contracts/ETHAssetHolder.sol]: 👍

File: contracts/ETHAssetHolder.sol (hunk 6 above)
Comment (geoknee) [reply contracts/ETHAssetHolder.sol]: See other comment.

File: test/ForceMove/_acceptableWhoSignedWhat.test.ts (hunk 9 above)
Comment (geoknee) [reply test/ForceMove/_acceptableWhoSignedWhat.test.ts]: 👍

File: contracts/Outcome.sol (hunk 8 above)
Comment (geoknee) [reply contracts/Outcome.sol]: OK

File: contracts/NitroAdjudicator.sol (hunk 7 above)
Comment (geoknee) [reply contracts/NitroAdjudicator.sol]: Right, this makes sense. Will start working on the fix 👍

File: test/ForceMove/_acceptableWhoSignedWhat.test.ts (hunk 9 above)
Comment (geoknee) [reply test/ForceMove/_acceptableWhoSignedWhat.test.ts]: Addresses in 0fc73f6

File: contracts/AssetHolder.sol (hunk 2 above)
Comment (geoknee) [reply contracts/AssetHolder.sol]: Addressed in 9b1fbc6

File: contracts/ETHAssetHolder.sol (hunk 5 above)
Comment (geoknee) [reply contracts/ETHAssetHolder.sol]: Addressed in 9b1fbc6

File: contracts/Outcome.sol (hunk 8 above)
Comment (geoknee) [reply contracts/Outcome.sol]: Addressed in 8439cf6

File: contracts/ERC20AssetHolder.sol (hunk 3 above)
Comment (geoknee) [reply contracts/ERC20AssetHolder.sol]: Have overwritten with your explanation (e7a7aaf)

File: contracts/ERC20AssetHolder.sol [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+contract ERC20AssetHolder is AssetHolder {
+    address AdjudicatorAddress;
+    address TokenAddress;
+
+    constructor(address _AdjudicatorAddress, address _TokenAddress) public {
+        AdjudicatorAddress = _AdjudicatorAddress;
+        TokenAddress = _TokenAddress;
+    }
+
+    modifier AdjudicatorOnly {
```
Comment (kerzhner) [inline contracts/ERC20AssetHolder.sol]: I might be missing something, but it looks like this modifier is not used.

File: contracts/Outcome.sol [hunk 11]
```diff
@@ -0,0 +1,37 @@
+pragma solidity ^0.5.11;
+pragma experimental ABIEncoderV2;
+
+library Outcome {
+    //An outcome is a set of asset outcomes.
+    // Outcome = AssetOutcome[]
```
Comment (kerzhner) [inline contracts/Outcome.sol]: This comment implies that an outcome can be a mix of allocations and guarantees. Is that the case? Or should an outcome only contain either allocations or guarantees?

File: contracts/NitroAdjudicator.sol [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+contract NitroAdjudicator is ForceMove {
+    function pushOutcome(
+        bytes32 channelId,
+        uint256 turnNumRecord,
+        uint256 finalizesAt,
+        bytes32 stateHash,
+        address challengerAddress,
+        bytes memory outcome,
```
Comment (kerzhner) [inline contracts/NitroAdjudicator.sol]: I wonder if there is a good way to define the `outcome` parameter as a struct instead of generic bytes.

File: contracts/ERC20AssetHolder.sol (hunk 10 above)
Comment (geoknee) [reply contracts/ERC20AssetHolder.sol]: It does appear that way; however, the modifier is used in the base contract `AssetHolder` that this contract extends. It is also defined there, but needs to be re-defined here in order to  create the constructor correctly with the `AdjudicatorAddress` (only known at deploy-time). 

Take a look at [this branch](https://github.com/statechannels/nitro-protocol/compare/asset-holder...adjudicator-only-modifier?expand=1) where I removed the modifier -- tests fail because the modifier is being defined only in the base contract, and hence before the `AdjudicatorAddress` is defined.

File: contracts/ETHAssetHolder.sol [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import './Outcome.sol';
+import './AssetHolder.sol';
+contract ETHAssetHolder is AssetHolder {
+    address AdjudicatorAddress;
+
+    constructor(address _AdjudicatorAddress) public {
+        AdjudicatorAddress = _AdjudicatorAddress;
+    }
+
+    modifier AdjudicatorOnly {
```
[the remark this answers is not in the document]
Comment (geoknee) [reply contracts/ETHAssetHolder.sol]: See comment above. Note that the constructor for this contract differs from that of `ERC20AssetHolder`, meaning that it cannot be defined in the base contract `AssetHolder`.

File: contracts/NitroAdjudicator.sol (hunk 7 above)
Comment (geoknee) [reply contracts/NitroAdjudicator.sol]: Addressed in 485324f

File: contracts/ERC20AssetHolder.sol [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        require(
+            holdings[destination] >= expectedHeld,
+            'Deposit: holdings[destination] is less than expected'
+        );
+
+        // If I expect there to be 10 and deposit 2, my goal was to get the
+        // balance to 12.
+        // In case some arbitrary person deposited 1 eth before I noticed, making the
+        // holdings 11, I should be refunded 1.
+        if (holdings[destination] == expectedHeld) {
```
[the remark this answers is not in the document]
Comment (geoknee) [reply contracts/ERC20AssetHolder.sol]: Yes -- I also realised the third branch could be folded into something more streamlined, too. Addressed in fe8a5d1

File: contracts/ERC20AssetHolder.sol [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        // 1. Participant A deposits
+        // 2. Participant B sees A's deposit, which means it is now safe for them to deposit
+        // 3. Participant B submits their deposit
+        // 4. The chain re-orgs, leaving B's deposit in the chain but not A's
+        require(
+            holdings[destination] >= expectedHeld,
+            'Deposit: holdings[destination] is less than expected'
+        );
+
+        // If I expect there to be 10 and deposit 2, my goal was to get the
```
[the remark this answers is not in the document]
Comment (geoknee) [reply contracts/ERC20AssetHolder.sol]: I rewrote this in the third person in fe8a5d1, and also used variable names in place of actual numbers.
